chain/rosetta-rpc/src/adapters/mod.rs

```diff
@@ -0,0 +1,1437 @@
+                    operations.push(
+                        validated_operations::AddKeyOperation {
+                            account: receiver_account_identifier.clone(),
+                            public_key: (&action.public_key).into(),
+                        }
+                        .into_related_operation(
+                            add_key_operation_id,
+                            vec![initiate_add_key_operation_id],
+                        ),
+                    );
+                }
+
+                near_primitives::transaction::Action::DeleteKey(action) => {
+                    let initiate_delete_key_operation_id =
+                        crate::models::OperationIdentifier::new(&operations);
+                    operations.push(
+                        validated_operations::InitiateDeleteKeyOperation {
+                            sender_account: sender_account_identifier.clone(),
+                        }
+                        .into_operation(initiate_delete_key_operation_id.clone()),
+                    );
+
+                    operations.push(
+                        validated_operations::DeleteKeyOperation {
+                            account: receiver_account_identifier.clone(),
+                            public_key: (&action.public_key).into(),
+                        }
+                        .into_related_operation(
+                            crate::models::OperationIdentifier::new(&operations),
+                            vec![initiate_delete_key_operation_id],
+                        ),
+                    );
+                }
+
+                near_primitives::transaction::Action::Transfer(action) => {
+                    let transfer_amount = crate::models::Amount::from_yoctonear(action.deposit);
+
+                    let sender_transfer_operation_id =
+                        crate::models::OperationIdentifier::new(&operations);
+                    operations.push(
+                        validated_operations::TransferOperation {
+                            account: sender_account_identifier.clone(),
+                            amount: -transfer_amount.clone(),
+                        }
+                        .into_operation(sender_transfer_operation_id.clone()),
+                    );
+
+                    operations.push(
+                        validated_operations::TransferOperation {
+                            account: receiver_account_identifier.clone(),
+                            amount: transfer_amount,
+                        }
+                        .into_related_operation(
+                            crate::models::OperationIdentifier::new(&operations),
+                            vec![sender_transfer_operation_id],
+                        ),
+                    );
+                }
+
+                near_primitives::transaction::Action::Stake(action) => {
+                    assert_eq!(sender_account_identifier, receiver_account_identifier);
```

Comment:
As discussed, the stake action can potentially happen on the different account, so long as the account didn't exist before and the first action of the batch was to create a new account. Then all the follow up actions are assuming they act on the given account.